ckan/lib/api_token.py

```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
+
+import jwt
+import logging
+
+from calendar import timegm
+
+import ckan.plugins as plugins
+import ckan.model as model
+from ckan.common import config
+from ckan.logic.schema import default_create_api_token_schema
+
+log = logging.getLogger(__name__)
+
+_config_encode_secret = u"api_token.jwt.encode.secret"
+_config_decode_secret = u"api_token.jwt.decode.secret"
+_config_secret_fallback = u"beaker.session.secret"
+
+_config_algorithm = u"api_token.jwt.algorithm"
+
+
+def _get_plugins():
+    return plugins.PluginImplementations(plugins.IApiToken)
+
+
+def _get_algorithm():
+    return config.get(_config_algorithm, u"HS256")
+
+
+def _get_secret(encode):
+    config_key = _config_encode_secret if encode else _config_decode_secret
+    secret = config.get(config_key)
+    if not secret:
+        secret = u'string:' + config.get(_config_secret_fallback)
+    type_, value = secret.split(u':', 1)
+    if type_ == u'file':
+        with open(value, u'rb') as key_file:
+            value = key_file.read()
+    if not value:
+        log.warning(
+            u"Neither `%s` nor `%s` specified. "
+            u"Missing secret key is a critical security issue.",
+            config_key,
+            _config_secret_fallback,
+        )
```

Comment:
Let's refuse to start ckan if there's no session secret instead of logging and continuing.

Reply:
ok. I hesitated whether it should be reported as an exception(personally I prefer to prevent any possibility of running instance in an obviously invalid state) or just notify that something is potentially broken.